test/test-util.ts

```diff
@@ -20,15 +20,132 @@ import {getTSConfig} from '../src/util';
 
 test('get should parse the correct tsconfig file', async t => {
   const FAKE_DIRECTORY = '/some/fake/directory';
-  const FAKE_CONFIG = {a: 'b'};
+  const FAKE_CONFIG1 = {files: ['b']};
   function fakeReadFilep(
       configPath: string, encoding: string): Promise<string> {
     t.is(configPath, path.join(FAKE_DIRECTORY, 'tsconfig.json'));
     t.is(encoding, 'utf8');
-    return Promise.resolve(JSON.stringify(FAKE_CONFIG));
+    return Promise.resolve(JSON.stringify(FAKE_CONFIG1));
   }
   const contents = await getTSConfig(FAKE_DIRECTORY, fakeReadFilep);
-  t.deepEqual(contents, FAKE_CONFIG);
+  t.deepEqual(contents, FAKE_CONFIG1);
 });
 
+test('should throw an error if it finds a circular reference', async t => {
+  const FAKE_DIRECTORY = '/some/fake/directory';
+  const FAKE_CONFIG1 = {files: ['b'], extends: 'FAKE_CONFIG2'};
+  const FAKE_CONFIG2 = {extends: 'FAKE_CONFIG3'};
+  const FAKE_CONFIG3 = {extends: 'FAKE_CONFIG1'};
+  function fakeReadFilep(
+      configPath: string, encoding: string): Promise<string> {
+    switch (configPath) {
+      case '/some/fake/directory/FAKE_CONFIG1':
+        return Promise.resolve(JSON.stringify(FAKE_CONFIG1));
+      case '/some/fake/directory/FAKE_CONFIG2':
+        return Promise.resolve(JSON.stringify(FAKE_CONFIG2));
+      case '/some/fake/directory/FAKE_CONFIG3':
+        return Promise.resolve(JSON.stringify(FAKE_CONFIG3));
+      case '/some/fake/directory/tsconfig.json':
+        return Promise.resolve(JSON.stringify(FAKE_CONFIG1));
+      default:
+        return Promise.reject('File Not Found');
+    }
+  }
+  await t.throws(
+      getTSConfig(FAKE_DIRECTORY, fakeReadFilep), Error,
+      'Circular Reference Detected');
+});
+
+test('should follow dependency chain caused by extends files', async t => {
+  const FAKE_DIRECTORY = '/some/fake/directory';
+  const FAKE_CONFIG1 = {
+    compilerOptions: {a: 'n'},
+    files: ['b'],
+    extends: 'FAKE_CONFIG2'
+  };
+  const FAKE_CONFIG2 = {include: ['/stuff/*'], extends: 'FAKE_CONFIG3'};
+  const FAKE_CONFIG3 = {exclude: ['doesnt/look/like/anything/to/me']};
+  const combinedConfig = {
+    compilerOptions: {a: 'n'},
+    files: ['b'],
+    include: ['/stuff/*'],
+    exclude: ['doesnt/look/like/anything/to/me']
+  };
+
+
+  function fakeReadFilep(
+      configPath: string, encoding: string): Promise<string> {
+    switch (configPath) {
+      case '/some/fake/directory/FAKE_CONFIG1':
+
```

Comment:
I think I missed this before. What is this case for? Is this falling through to the next case?

(Here and also in the next test)

Reply:
Similar to a symlink the file FAKE_CONFIG1 has 2 addresses /some/fake/directory/tsconfig.json and /some/fake/directory/FAKE_CONFIG1. I could edit this out and make the circular references point back to /some/fake/directory/tsconfig.json I kept this in all the tests for consistancy but I can remove it